### PR TITLE
optimize array and struct construction from APPEND and SETITEM to PACK

### PR DESCRIPTION
**Pull request by Hecate2**
Saving GAS. APPEND costs 8192 each, while PACK costs 2048.
Will conflict. draft for now.

### CODE REVIEW

**Comment by Hecate2** [comment]
Will conflict in `UnitTest_DebugInfo.cs`. Requires manual updates in tests.

**Comment by Jim8y** [comment]
@shargon This pr is awesome, it optimized some gas consumption by 10%-50%.

File: src/Neo.Compiler.CSharp/MethodConvert/Expression/AnonymousObjectCreationExpression.cs [hunk 1]
```diff
@@ -35,12 +36,9 @@ internal partial class MethodConvert
     /// <seealso href="https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/types/anonymous-types">Anonymous types</seealso>
     private void ConvertAnonymousObjectCreationExpression(SemanticModel model, AnonymousObjectCreationExpressionSyntax expression)
     {
-        AddInstruction(OpCode.NEWARRAY0);
-        foreach (AnonymousObjectMemberDeclaratorSyntax initializer in expression.Initializers)
-        {
-            AddInstruction(OpCode.DUP);
+        foreach (AnonymousObjectMemberDeclaratorSyntax initializer in expression.Initializers.Reverse())
```
Comment (shargon) [inline src/Neo.Compiler.CSharp/MethodConvert/Expression/AnonymousObjectCreationExpression.cs:39]:
```suggestion
        foreach (AnonymousObjectMemberDeclaratorSyntax initializer in expression.Initializers.Reverse())  // PACK works in a reversed way
```

File: src/Neo.Compiler.CSharp/MethodConvert/Expression/TupleExpression.cs [hunk 2]
```diff
@@ -37,12 +38,9 @@ internal partial class MethodConvert
     /// <seealso href="https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/value-tuples">Tuple types</seealso>
     private void ConvertTupleExpression(SemanticModel model, TupleExpressionSyntax expression)
     {
-        AddInstruction(OpCode.NEWSTRUCT0);
-        foreach (ArgumentSyntax argument in expression.Arguments)
-        {
-            AddInstruction(OpCode.DUP);
+        foreach (ArgumentSyntax argument in expression.Arguments.Reverse())
```
Comment (shargon) [inline src/Neo.Compiler.CSharp/MethodConvert/Expression/TupleExpression.cs:41]:
```suggestion
        foreach (ArgumentSyntax argument in expression.Arguments.Reverse())  // PACKSTRUCT works in a reversed way
```

File: src/Neo.Compiler.CSharp/MethodConvert/Helpers/ConvertHelpers.cs [hunk 3]
```diff
@@ -209,28 +209,25 @@ private void CreateObject(SemanticModel model, ITypeSymbol type, InitializerExpr
...
+            return;
         }
-        var virtualMethods = members.OfType<IMethodSymbol>().Where(p => p.IsVirtualMethod()).ToArray();
-        if (type.IsRecord || virtualMethods.Length <= 0) return;
-        var index = _context.AddVTable(type);
-        AddInstruction(OpCode.DUP);
-        AccessSlot(OpCode.LDSFLD, index);
-        AddInstruction(OpCode.APPEND);
+
+        foreach (var field in fields.Reverse())
```
Comment (shargon) [inline src/Neo.Compiler.CSharp/MethodConvert/Helpers/ConvertHelpers.cs:228]:
```suggestion
        foreach (var field in fields.Reverse())  // PACKS and PACKSTRUCT works in a reversed way
```

File: src/Neo.Compiler.CSharp/MethodConvert/System/SystemCall.Enum.cs [hunk 4]
```diff
@@ -518,16 +518,10 @@ private static void HandleEnumGetNames(MethodConvert methodConvert, SemanticMode
...
-        methodConvert.AddInstruction(OpCode.NEWARRAY); // Stack: [..., array]
-
-        for (int i = 0; i < enumMembers.Length; i++)
-        {
-            methodConvert.Dup();                      // Duplicate array reference
-            methodConvert.Push(i);                    // Index
-            methodConvert.Push(enumMembers[i].Name);  // Name
-            methodConvert.AddInstruction(OpCode.SETITEM); // array[i] = name
-        }
+        foreach (IFieldSymbol m in enumMembers.Reverse())
```
Comment (shargon) [inline src/Neo.Compiler.CSharp/MethodConvert/System/SystemCall.Enum.cs:521]:
```suggestion
        foreach (IFieldSymbol m in enumMembers.Reverse())  // PACK works in a reversed way
```

File: src/Neo.Compiler.CSharp/MethodConvert/System/SystemCall.Enum.cs [hunk 5]
```diff
@@ -568,16 +562,10 @@ private static void HandleEnumGetValues(MethodConvert methodConvert, SemanticMod
...
-        methodConvert.AddInstruction(OpCode.NEWARRAY); // Stack: [..., array]
-
-        for (int i = 0; i < enumMembers.Length; i++)
-        {
-            methodConvert.Dup();                      // Duplicate array reference
-            methodConvert.Push(i);                    // Index
-            methodConvert.Push(enumMembers[i].ConstantValue); // Value
-            methodConvert.AddInstruction(OpCode.SETITEM); // array[i] = value
-        }
+        foreach (IFieldSymbol m in enumMembers.Reverse())
```
Comment (shargon) [inline src/Neo.Compiler.CSharp/MethodConvert/System/SystemCall.Enum.cs:565]:
```suggestion
        foreach (IFieldSymbol m in enumMembers.Reverse())  // PACK works in a reversed way
```

**Review by shargon — APPROVED** [review]
Amazing improve 👏

**Pull request merged**